### PR TITLE
RGW - Add POSIX Driver

### PR DESCRIPTION
**Pull request by dang**
This is the MVP for a driver for RGW that operates on top of a POSIX filesystem.  It supports get, put, list, copy, multipart, external access via the filesystem itself, and ordered bucket listings via an LRU-based cache.

## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "pacific"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

## Checklist
- Tracker (select at least one)
  - [ ] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [x] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [x] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [x] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [x] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
</details>

### CODE REVIEW

File: src/rgw/driver/posix/rgw_sal_posix.cc
Comment (cbodley) [inline src/rgw/driver/posix/rgw_sal_posix.cc:1495]: how are intermediate errors communicated back to the caller?

File: src/rgw/driver/posix/rgw_sal_posix.h
Comment (cbodley) [inline src/rgw/driver/posix/rgw_sal_posix.h:191]: copy constructor should take const ref (same for POSIXObject)

File: src/rgw/driver/posix/rgw_sal_posix.h
Comment (cbodley) [inline src/rgw/driver/posix/rgw_sal_posix.h:205]: isn't `info` a member of `StoreBucket`? why doesn't its `set_owner()` handle this?

File: src/rgw/driver/posix/rgw_sal_posix.cc
Comment (cbodley) [inline src/rgw/driver/posix/rgw_sal_posix.cc:455]: this variable look unused?

**Review by cbodley — COMMENTED** [review]
a lot to review, but looks wonderful overall

File: src/rgw/driver/posix/lmdb-safe.cc
Comment (mattbenjamin) [reply src/rgw/driver/posix/lmdb-safe.cc:13]: yes, ok--these are mit/bsd, will source

File: src/rgw/driver/posix/bucket_cache.h
Comment (mattbenjamin) [reply src/rgw/driver/posix/bucket_cache.h:160]: well, what I was doing here was a) preferring to use normal_link; but b) have this to print info

File: src/common/options/rgw.yaml.in
Comment (thotz) [inline src/common/options/rgw.yaml.in:3621]: part of rgw_config_store?

File: src/rgw/driver/posix/lmdb-safe.cc
Comment (dang) [reply src/rgw/driver/posix/lmdb-safe.cc:13]: @mattbenjamin ?

File: src/rgw/rgw_op.cc
Comment (dang) [reply src/rgw/rgw_op.cc:189]: To fix a s3tests failure.  I can revert for now, if you want, but it'll quite possibly come back when I get back to working on s3-tests.

File: src/rgw/driver/posix/CMakeLists.txt
Comment (dang) [reply src/rgw/driver/posix/CMakeLists.txt:25]: quite possibly.

File: src/rgw/rgw_sal_filter.h
Comment (dang) [reply src/rgw/rgw_sal_filter.h:19]: Not sure.  It doesn't seem to matter.

File: src/rgw/driver/posix/rgw_sal_posix.cc
Comment (dang) [reply src/rgw/driver/posix/rgw_sal_posix.cc:31]: Left over from early work.  will fix.

File: src/rgw/driver/posix/rgw_sal_posix.cc
Comment (dang) [reply src/rgw/driver/posix/rgw_sal_posix.cc:668]: Using a unique_ptr requires jumping through hoops when assigning it to the out parameter, since it's a POSIXBucket not a Bucket.  Ugly either way, so I don't have a huge preference.

File: src/rgw/driver/posix/rgw_sal_posix.h
Comment (dang) [reply src/rgw/driver/posix/rgw_sal_posix.h:282]: Okay.

File: src/rgw/rgw_op.cc
Comment (cbodley) [reply src/rgw/rgw_op.cc:189]: i'd like to understand that failure better. you duplicate `decode_policy()` in rgw_sal_polix.cc with the EIO error code, but changed the common code that applies to other drivers which didn't fail here

File: src/rgw/driver/posix/rgw_sal_posix.h
Comment (dang) [reply src/rgw/driver/posix/rgw_sal_posix.h:684]: Left over from interim work.

File: src/rgw/driver/posix/rgw_sal_posix.cc
Comment (dang) [reply src/rgw/driver/posix/rgw_sal_posix.cc:2913]: Those are probably cases where the yield was added in the rebase on main, but didn't exist when implemented.  I'll look.

File: src/rgw/driver/posix/rgw_sal_posix.cc
Comment (dang) [reply src/rgw/driver/posix/rgw_sal_posix.cc:2190]: etag seems to be will defined as the md5sum of the contents of the file, so no.  The plan is to eventually add a background service based on fnotify that catches adds and generates etag.

File: src/rgw/driver/posix/rgw_sal_posix.h
Comment (dang) [reply src/rgw/driver/posix/rgw_sal_posix.h:280]: I did it this way to avoid code duplication.

File: src/rgw/driver/posix/rgw_sal_posix.cc
Comment (dang) [reply src/rgw/driver/posix/rgw_sal_posix.cc:1495]: will fix.

File: src/rgw/driver/posix/rgw_sal_posix.h
Comment (dang) [reply src/rgw/driver/posix/rgw_sal_posix.h:191]: Okay

File: src/rgw/driver/posix/rgw_sal_posix.h
Comment (dang) [reply src/rgw/driver/posix/rgw_sal_posix.h:205]: good question

File: src/rgw/driver/posix/rgw_sal_posix.cc
Comment (cbodley) [reply src/rgw/driver/posix/rgw_sal_posix.cc:668]: unique_ptr move-assignment from a derived type to a base type works fine. this is overload (2) in https://en.cppreference.com/w/cpp/memory/unique_ptr/operator%3D

i wrote a trivial example in https://godbolt.org/z/rf378cvqE

File: src/rgw/driver/posix/rgw_sal_posix.cc
Comment (dang) [reply src/rgw/driver/posix/rgw_sal_posix.cc:455]: Yep

File: src/common/options/rgw.yaml.in
Comment (dang) [reply src/common/options/rgw.yaml.in:3621]: I don't understand the question?  In the future (work Ali is working on) all this will be replaced with JSON stored in the zonegroup, but for now it has to live in ceph.conf.

File: src/rgw/driver/posix/rgw_sal_posix.h
Comment (cbodley) [reply src/rgw/driver/posix/rgw_sal_posix.h:280]: understood, but you can share code without sharing the interface. i'm just picturing a `class POSIXBucket` and `class MultipartDirectory` that both call helper functions that operate on file descriptors and paths

with this separation, it would be much easier to make changes or bugfixes to one without introducing weird regressions in the other

File: src/common/options/rgw.yaml.in
Comment (cbodley) [reply src/common/options/rgw.yaml.in:3621]: i don't think it's necessary to have a posix-file backend for `rgw::sal::ConfigStore`. `rgw_config_store=dbstore` already works on a local filesystem

File: src/rgw/rgw_op.cc
Comment (dang) [reply src/rgw/rgw_op.cc:189]: I'll back this out until I can reproduce and find which test was failing.

File: src/rgw/driver/posix/rgw_sal_posix.cc
Comment (dang) [reply src/rgw/driver/posix/rgw_sal_posix.cc:668]: Interesting.  It certainly didn't work on the compiler I had when I wrote RADOSStore.  That's why I used this pattern.

File: src/common/options/rgw.yaml.in
Comment (mattbenjamin) [reply src/common/options/rgw.yaml.in:3621]: is "JSON stored in the zonegroup" equivalent to putting things in sqlite?  I don't follow

File: src/rgw/driver/posix/bucket_cache.h
Comment (mattbenjamin) [reply src/rgw/driver/posix/bucket_cache.h:160]: it's an #if iow because I'd like this normally to have normal_link; I can templatize this or something later.  I don't want asserts or ceph_asserts in the path.  if we're going to change this now, make it normal_link in the decl and just remove this block

File: src/common/options/rgw.yaml.in
Comment (dang) [reply src/common/options/rgw.yaml.in:3621]: > is "JSON stored in the zonegroup" equivalent to creating a posix-file backend for rgw::sal::ConfigStore? I don't follow

No, it describes the driver stack, so it's a replacement for `rgw_backend_store` and `rgw_filter` currently in ceph.conf.  The actual zonegroup is stored in the ConfigStore, so whichever ConfigStore you use will store the JSON that describes the driver stack.  I agree with Casey, we'll probably use DBStore's ConfigStore for POSIXDriver.

Comment (dang) [reply src/common/options/rgw.yaml.in:3621]: > part of rgw_config_store?

Ah, was the original question asking if this should be in rgw_config_store instead of rgw_filter?  If so, then yes, eventually.  Currently, POSIXDriver is a filter, since it doesn't implement any User APIs, but instead depends on DBUser.  So it has to be a filter stacked on top of DBStore until we finish the User APIs.

File: src/rgw/driver/posix/rgw_sal_posix.cc
Comment (cbodley) [reply src/rgw/driver/posix/rgw_sal_posix.cc:2188]: this issue doesn't need to block merge, but i'd really like to see further discussion here. this feels like a major scalability issue when compared to other backends, and will be very visible to s3 clients as request timeouts and unpredictable latency

> etag seems to be will defined as the md5sum of the contents of the file, so no

it is defined, but with special cases for multipart and encrypted objects. https://docs.aws.amazon.com/AmazonS3/latest/userguide/checking-object-integrity.html#checking-object-integrity-etag-and-md5 is a good resource on this

we should consider defining a new special case for the posix backend to avoid this requirement to read a potentially-unbounded amount of data

some s3 clients like s3cmd try to use the etag for checksum verification, but it should generally be possible to disable that with options like `--no-check-md5`. there was a related discussion in https://github.com/s3tools/s3cmd/issues/880, where it looks like minio worked around a similar issue by adding a `-1` suffix so the etag would be interpreted as multipart

aside from the s3 protocol, http has its own requirements for [ETag](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/ETag) related to its use in caching and the `If-Match`/`If-None-Match` headers. there i believe the only requirement is that the etag changes if the data changes, so some combination of the inode number and mtime may work as a proxy for the md5sum

> The plan is to eventually add a background service based on fnotify that catches adds and generates etag.

i can see how this would reduce the probability that a client reads an object with uninitialized etag so needs to calculate the md5sum inline. but it still imposes a lot of background i/o that shouldn't be necessary

i guess that would also have to cover filesystem writes to existing objects that already have an etag? what if there's an append-only workload, like a log file? this background worker would have to re-read the entire file for checksumming often, yet its actual etag would rarely be consistent

Comment (dang) [reply src/rgw/driver/posix/rgw_sal_posix.cc:2188]: I don't have a strong opinion about this.  I added it so that s3cmd would succeed.  It doesn't matter for passing things like s3-tests, since those will never deal with side-loaded data.   If things other than s3cmd really don't check etag in general, then I'm fine with just removing it, and never having an etag set.  The way it's written now is certainly a bandaid, and can't handle may cases, including overwriting or appending outside of RGW, so it can't be a long-term solution.   My impression was that etag was important, and everything used it to verify on every get, but I've never verified that.

**Comment by cbodley** [comment]
jenkins test windows

File: src/test/rgw/CMakeLists.txt
Comment (cbodley) [inline src/test/rgw/CMakeLists.txt:327]: there should be a `add_ceph_unittest()` to make this run during 'make check'

**Review by cbodley — APPROVED** [review]
good to merge once we see `unittest_posix_bucket_cache` passing in 'make check'

**Comment by cbodley** [comment]
> 195/286 Test #232: unittest_posix_bucket_cache ...............***Failed   17.59 sec
...
[==========] 23 tests from 1 test suite ran. (17392 ms total)
[  PASSED  ] 20 tests.
[  FAILED  ] 3 tests, listed below:
[  FAILED  ] BucketCache.ListMarker1
[  FAILED  ] BucketCache.ListInotify1
[  FAILED  ] BucketCache.List2Inotify1

**Comment by dang** [comment]
> > 195/286 Test #232: unittest_posix_bucket_cache ...............***Failed   17.59 sec
> > ...
> > [==========] 23 tests from 1 test suite ran. (17392 ms total)
> > [  PASSED  ] 20 tests.
> > [  FAILED  ] 3 tests, listed below:
> > [  FAILED  ] BucketCache.ListMarker1
> > [  FAILED  ] BucketCache.ListInotify1
> > [  FAILED  ] BucketCache.List2Inotify1

Yep, I'm looking at it.

**Comment by dang** [comment]
jenkins test make check

File: src/rgw/driver/posix/rgw_sal_posix.cc
Comment (mattbenjamin) [inline src/rgw/driver/posix/rgw_sal_posix.cc:747]: looks good

**Review by mattbenjamin — APPROVED** [review]
lgtm (mint_listing_entry)

**Comment by mattbenjamin** [comment]
yay

**Pull request merged**
